eth/eventhandler/handlers.go

```diff
@@ -79,11 +80,46 @@ func (eh *EventHandler) handleOperatorAdded(txn basedb.Txn, event *contract.Cont
 	if err != nil {
 		return fmt.Errorf("save operator data: %w", err)
 	}
+
 	if exists {
 		logger.Debug("operator data already exists")
 		return nil
 	}
 
+	var modifiedShares []*ssvtypes.SSVShare
+	for _, share := range eh.nodeStorage.Shares().List(txn, registrystorage.ByOperatorID(event.OperatorId)) {
+		if !share.Liquidated {
+			continue
+		}
```

Comment:
We also have the ClusterLiquidated event which may make a share `Liquidated`, so it can be both true and false

Reply:
Right, perhaps leaving a comment noting that wouldn't hurt :)

But then I wonder what happens to **validator** in the following scenario:
- 4 operators are running validator
- at some point validator(shares) gets liquidated because he couldn't pay his bills
- then those 4 operators themselves get de-registered from SSV network (doesn't need to happen at the same time, could be many days of delay between each de-registering)
- then those 4 operators re-register again - also reviving this validator (that shouldn't  ever get revived in that manner) ?

well, actually the whole validator won't get revived ^ if I'm reading the code correctly, only the share belonging to the operator referenced in the event will be revived (unliquidated) ... which also means (again, if I got that right) this whole algorithm will only work as expected to handle just 1 operator de-regestering -> re-regestering (won't work in scenario where 2 operators out of N de-register, and only then after that try to re-register again)

Reply:
@iurii-ssv I agree about the comment. But I'm not sure I understand your concern about the re-registering logic. We process events one by one, so if 2 operators exist and 2 are trying to re-register, then after processing the first one there would be 3 active operators, so registering the second operator would lead to 4 active operators and share reactivation. If I didn't catch your thought correctly, please follow-up

Reply:
Well yeah, this operator/share management logic is a bit more involved than I expected ... so I decided to write a test or two to verify whether I understand it correctly ... and got a bit stuck.

There seems to be no "combined" tests that would define interaction between operator and his shares - but rather these aspects are tested in isolation. Here is my sketch for how they could look like (essentially turning my comment from above into code, and adding some more) - [a branch](https://github.com/ssvlabs/ssv/tree/testing/allow-creating-removed-operator) with 1 commit on top of what you have in this PR's branch - see [diff](https://github.com/ssvlabs/ssv/compare/allow-creating-removed-operator...testing/allow-creating-removed-operator). Note also, there are 2 minor fixes in there too, namely (so I recommend to "take as is and adjust it as you see fit"):
- `sharePubKey := validatorData2.operatorsShares[0].sec.GetPublicKey().Serialize()` seems to have been a typo
- and what I mentioned in https://github.com/ssvlabs/ssv/pull/1796#discussion_r1825720014
- plus some minor adjustments to `test ClusterLiquidated + ClusterReactivated events handling`

why did I get stuck ... well, first of all I don't really know what behavior we want to have (to define proper expectations in tests), I did it best I could and encountered that we [increment `operatorId` (in smart contract)](https://github.com/ssvlabs/ssv/blob/stage/eth/simulator/simcontract/simcontract.sol#L53) even for operators who are re-registering (using same pubkey - they end up with different `operatorId`)
- as the result ^ we can't utilize `operatorId` for certain things, for example [when checking whether share belongs to operator or not](https://github.com/ssvlabs/ssv/blob/7e058e953dbf7299ceaa595564819d95fffa7d6a/eth/eventhandler/handlers.go#L427-L440) - meaning, re-registered operator can't find his "old" share
- hence the tests I've added cannot digest that, but I don't think they are the problem; rather it seems we should either not increment `operatorId` in smart contract or not rely/use it like we do
- also, less relevant, but I'm not sure why we only have a part of solidity code but not the rest (the rest I guess is pre-compiled and used in tests ?), and hence can't quite understand how these events compare to what we have in main smart contract

Reply:
@iurii-ssv I think that contract is just for simulator, the actual contract is likely this one https://github.com/ssvlabs/ssv-network/blob/583b7b7cb1c1abc5d4c3b13bafca59bf315113b6/contracts/modules/SSVOperators.sol#L26

Can you please create a PR for your changes?

Reply:
> Can you please create a PR for your changes?

Yup, why not - https://github.com/ssvlabs/ssv/pull/1842